bokeh/io.py

```diff
@@ -424,6 +431,67 @@ def push(session_id=None, url=None, app_path=None, document=None, state=None, io
     _push_to_server(session_id=session_id, url=url, app_path=app_path,
                     document=document, io_loop=io_loop)
 
+def push_notebook(document=None, state=None):
+    ''' Update the last-shown plot in a Jupyter notebook with the new data
+    or property values.
+
+    Args:
+
+        document (Document, optional) :
+            A :class:`~bokeh.document.Document` to push from. If None,
+            uses ``curdoc()``.
+
+        state (State, optional) :
+            A Bokeh State object
+
+    Returns:
+        None
+
+    Examples:
+
+        Typical usage is typically similar to this:
+
+        .. code-block:: python
+
+            from bokeh.io import push_notebook
+
+            # code to create a plot
+
+            show(plot)
+
+            plot.title = "New Title"
+
+            # This will cause the title to update
+            push_notebook()
+
+    '''
+    if state is None:
+        state = _state
+
+    if state.server_enabled:
+        raise RuntimeError("output_server() has been called, use push() to push to server")
+
+    if not document:
+        document = state.document
+
+    if not document:
+        warnings.warn("No document to push")
+        return
+
+    if not state.last_comms:
+        warnings.warn("Cannot find a last shown plot to update. Call output_notebook() and show() before push_notebook()")
+        return
+
+    import json
+    to_json = document.to_json()
+    if state.last_pushed_doc and state.last_pushed_doc != document:
+        msg = dict(doc=to_json)
+    else:
+        msg = Document._compute_patch_between_json(state.last_pushed_json, to_json, document)
```

Comment:
I think these hand-constructed messages don't have the msgtype or msgid, which is fine since they don't go over the websocket, except how will the coffee tell whether it's a PUSH-DOC or a PATCH-DOC?